rfcs/20230801-max-partition-policy/README.md

```diff
@@ -0,0 +1,116 @@
+# Introducing Max Partition Policy to oneDNN Graph API
+
+## Motivation
+
+oneDNN Graph Compiler is currently a experimental backend for oneDNN Graph API.
+It provides
+optimized implementations for complex computational graphs including multi-head
+attention (MHA), multi-layer perceptron (MLP), and convolution residual blocks
+over typical data types for both inference and training.
+
+To further boost performance and expose more aggressive optimizations to users,
+oneDNN Graph Compiler is requesting to add max partition
+policy to oneDNN Graph API.
+
+## Current Status
+
+In oneDNN Graph API programming model, a computation graph is passed to library
+and then optimized partitions are returned by the library by
+`graph.get_partitions()` API.
+
+```c
+dnnl_status_t dnnl_graph_graph_filter(
+        dnnl_graph_graph_t graph, dnnl_graph_partition_policy_t policy);
+```
+
+```cpp
+class graph {
+    ...
+    std::vector<partition> get_partitions(
+            partition::policy policy = partition::policy::fusion);
+    ...
+};
+```
+
+A partition policy is used in the API to control granularities
+of optimizations to be applied when partitioning.
+Currently, oneDNN Graph API supports two partition policies:
+
+```c
+typedef enum {
+    dnnl_graph_partition_policy_fusion = 1,
+    dnnl_graph_partition_policy_debug = 2,
+} dnnl_graph_partition_policy_t;
+```
+
+```cpp
+class partition {
+    ...
+    enum class policy {
+        fusion = dnnl_graph_partition_policy_fusion,
+        debug = dnnl_graph_partition_policy_debug,
+    };
+    ...
+};
+```
+
+- Fusion policy (default partition policy): fusion policy returns partitions
+  with [pre-defined fusion patterns](https://oneapi-src.github.io/oneDNN/dev_guide_graph_fusion_patterns.html),
+  e.g. Convolution followed by a chain of post-ops, MHA, MLP, etc.
+- Debug policy: debug policy returns partitions with single operation in each partition.
+  This policy doesn't apply any fusions, which is useful when
+  user wants to debug or compare performance with fusion policy.
+
+## Proposal
+
+The API to expose max policy to users is similar to fusion and debug policies.
+
+```c
+typedef enum {
+    dnnl_graph_partition_policy_max = 0, // the new max policy
+    dnnl_graph_partition_policy_fusion = 1,
+    dnnl_graph_partition_policy_debug = 2,
+} dnnl_graph_partition_policy_t;
+```
+
+```cpp
+class partition {
+    ...
+    enum class policy {
+        max = dnnl_graph_partition_policy_max, // the new max policy
+        fusion = dnnl_graph_partition_policy_fusion,
+        debug = dnnl_graph_partition_policy_debug,
+    };
+    ...
+};
+```
+
+The expected behaviors when using max policy:
+
+- If not specified, the default partition policy remains to be fusion policy.
+- Max policy returns partitions max in partition size. The criteria of partitioning
```

Comment:
What is motivation for differentiation between max and fusion? How user can decide which one should be used (not for a single model, but for overall integration)? Maybe "max" should be renamed to something like "experimental" ("experimental" is really bad word, we already have semantics for experimental features) to indicate that user shouldn't use it until they know it brings performance for their particular case?

Reply:
> What is motivation for differentiation between max and fusion?

From my understanding, fusion defines fixed number of patterns, while max's optimizations are not fixed, it will generate different optimization for different models based on model architecture, which might be good for OOB models. 

> How user can decide which one should be used (not for a single model, but for overall integration)?

It depends on how much control user wants to give to the library. Compared with fusion, max gives the library more control to optimize the graph.
But I can understand user wants to get better performance when they give the library more control, which is not guaranteed for max policy since we are not able to test all the models in the world (Please correct me if I'm wrong @ZhennanQin @niuxiaog).

> Maybe "max" should be renamed to something like "experimental" ("experimental" is really bad word, we already have semantics for experimental features) to indicate that user shouldn't use it until they know it brings performance for their particular case? Another option to call it "all" if the intend is to take all graph and merge it into a single partition.

`all` may not be true because `take all graph and merge it into a single partition` is not guaranteed. `experimental` sounds a doable option, what do you think? @ZhennanQin @TaoLv

Reply:
`max` policy is designed to support direct programming with oneDNN graph API to implement custom ops or custom fusion patterns. For example, end-user can build RMS_norm with small ops and pass it to onednn graph with max policy, and compiler backend can generate a fused kernel for this. This is frequent request for newly introduced activation or normalization. I think `experimental` is not a good name since it can mean anything and doesn't reflect actual behavior. But we can use other term than `max` to make it more clear.